### PR TITLE
[DROOLS-1175] infer numeric type for sum expression in an accumulate …

### PR DESCRIPTION
**Pull request by mariofusco**
…pattern

### CODE REVIEW

**Comment by ge0ffrey** [comment]
All optaplanner's non-turtle tests succeed, but this DRL fails:

```
rule "requiredCpuPowerTotal"
    when
        $computer : CloudComputer($cpuPower : cpuPower)
        $requiredCpuPowerTotal : Integer(intValue > $cpuPower) from accumulate(
            CloudProcess(
                computer == $computer,
                $requiredCpuPower : requiredCpuPower),
            sum($requiredCpuPower)
        )
    then
        scoreHolder.addHardConstraintMatch(kcontext, $cpuPower - $requiredCpuPowerTotal);
end
```

with this exception:

```
Caused by: Exception executing consequence for rule "requiredCpuPowerTotal" in org.optaplanner.examples.cloudbalancing.solver: java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Integer
	at org.drools.core.runtime.rule.impl.DefaultConsequenceExceptionHandler.handleException(DefaultConsequenceExceptionHandler.java:39)
	at org.drools.core.common.DefaultAgenda.fireActivation(DefaultAgenda.java:1100)
	at org.drools.core.phreak.RuleExecutor.fire(RuleExecutor.java:121)
	at org.drools.core.phreak.RuleExecutor.evaluateNetworkAndFire(RuleExecutor.java:74)
	at org.drools.core.common.DefaultAgenda.fireNextItem(DefaultAgenda.java:1007)
	at org.drools.core.common.DefaultAgenda.fireLoop(DefaultAgenda.java:1351)
	at org.drools.core.common.DefaultAgenda.fireAllRules(DefaultAgenda.java:1289)
	at org.drools.core.impl.StatefulKnowledgeSessionImpl.internalFireAllRules(StatefulKnowledgeSessionImpl.java:1335)
	at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1326)
	at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1307)
	at org.optaplanner.core.impl.score.director.drools.DroolsScoreDirector.calculateScore(DroolsScoreDirector.java:86)
	at org.optaplanner.core.impl.solver.scope.DefaultSolverScope.calculateScore(DefaultSolverScope.java:109)
	at org.optaplanner.core.impl.phase.scope.AbstractPhaseScope.calculateScore(AbstractPhaseScope.java:159)
	at org.optaplanner.core.impl.constructionheuristic.decider.ConstructionHeuristicDecider.processMove(ConstructionHeuristicDecider.java:133)
	at org.optaplanner.core.impl.constructionheuristic.decider.ConstructionHeuristicDecider.doMove(ConstructionHeuristicDecider.java:122)
	at org.optaplanner.core.impl.constructionheuristic.decider.ConstructionHeuristicDecider.decideNextStep(ConstructionHeuristicDecider.java:95)
	at org.optaplanner.core.impl.constructionheuristic.DefaultConstructionHeuristicPhase.solve(DefaultConstructionHeuristicPhase.java:64)
	at org.optaplanner.core.impl.solver.DefaultSolver.runPhases(DefaultSolver.java:237)
	at org.optaplanner.core.impl.solver.DefaultSolver.solve(DefaultSolver.java:198)
```

**Comment by ge0ffrey** [comment]
This rule:

> rule "requiredCpuPowerTotal"
>     when
>         $computer : CloudComputer($cpuPower : cpuPower)
>         $requiredCpuPowerTotal : Integer(intValue > $cpuPower) from accumulate(
>             CloudProcess(
>                 computer == $computer,
>                 $requiredCpuPower : requiredCpuPower),
>             sum($requiredCpuPower)
>         )
>     then
>         scoreHolder.addHardConstraintMatch(kcontext, $cpuPower - $requiredCpuPowerTotal);
> end

fails with

> Caused by: Exception executing consequence for rule "requiredCpuPowerTotal" in org.optaplanner.examples.cloudbalancing.solver: java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Integer
> 	at org.drools.core.runtime.rule.impl.DefaultConsequenceExceptionHandler.handleException(DefaultConsequenceExceptionHandler.java:39)
> 	at org.drools.core.common.DefaultAgenda.fireActivation(DefaultAgenda.java:1100)
> 	at org.drools.core.phreak.RuleExecutor.fire(RuleExecutor.java:121)
> 	at org.drools.core.phreak.RuleExecutor.evaluateNetworkAndFire(RuleExecutor.java:74)
> 	at org.drools.core.common.DefaultAgenda.fireNextItem(DefaultAgenda.java:1007)
> 	at org.drools.core.common.DefaultAgenda.fireLoop(DefaultAgenda.java:1351)
> 	at org.drools.core.common.DefaultAgenda.fireAllRules(DefaultAgenda.java:1289)
> 	at org.drools.core.impl.StatefulKnowledgeSessionImpl.internalFireAllRules(StatefulKnowledgeSessionImpl.java:1335)
> 	at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1326)
> 	at org.drools.core.impl.StatefulKnowledgeSessionImpl.fireAllRules(StatefulKnowledgeSessionImpl.java:1307)
> 	at org.optaplanner.core.impl.score.director.drools.DroolsScoreDirector.calculateScore(DroolsScoreDirector.java:86)
> 	at org.optaplanner.core.impl.solver.scope.DefaultSolverScope.calculateScore(DefaultSolverScope.java:109)
> 	at org.optaplanner.core.impl.phase.scope.AbstractPhaseScope.calculateScore(AbstractPhaseScope.java:159)
> 	at org.optaplanner.core.impl.constructionheuristic.decider.ConstructionHeuristicDecider.processMove(ConstructionHeuristicDecider.java:133)
> 	at org.optaplanner.core.impl.constructionheuristic.decider.ConstructionHeuristicDecider.doMove(ConstructionHeuristicDecider.java:122)
> 	at org.optaplanner.core.impl.constructionheuristic.decider.ConstructionHeuristicDecider.decideNextStep(ConstructionHeuristicDecider.java:95)
> 	at org.optaplanner.core.impl.constructionheuristic.DefaultConstructionHeuristicPhase.solve(DefaultConstructionHeuristicPhase.java:64)
> 	at org.optaplanner.core.impl.solver.DefaultSolver.runPhases(DefaultSolver.java:237)
> 	at org.optaplanner.core.impl.solver.DefaultSolver.solve(DefaultSolver.java:198)
> 	at org.optaplanner.examples.common.business.SolutionBusiness.solve(SolutionBusiness.java:306)
> 	at org.optaplanner.examples.common.swingui.SolverAndPersistenceFrame$SolveWorker.doInBackground(SolverAndPersistenceFrame.java:311)
> 	at javax.swing.SwingWorker$1.call(SwingWorker.java:295)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at javax.swing.SwingWorker.run(SwingWorker.java:334)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
> 	at java.lang.Thread.run(Thread.java:745)
> Caused by: java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Integer
> 	at org.optaplanner.examples.cloudbalancing.solver.Rule_requiredCpuPowerTotal1027060274DefaultConsequenceInvokerGenerated.evaluate(Unknown Source)
> 	at org.optaplanner.examples.cloudbalancing.solver.Rule_requiredCpuPowerTotal1027060274DefaultConsequenceInvoker.evaluate(Unknown Source)
> 	at org.drools.core.common.DefaultAgenda.fireActivation(DefaultAgenda.java:1089)
> 	... 25 more
> 

It fails against master too, but I 'd expected it to work with this change.

**Comment by ge0ffrey** [comment]
I just figured out why: there's a BigDecimalSumAccumulateFunction, but no IntSumAccumulateFunction or LongSumAccumulateFunction yet. I 'll try implementing those latter and send you a PR for those :)

File: drools-compiler/src/main/java/org/drools/compiler/rule/builder/dialect/java/JavaAccumulateBuilder.java [hunk 1]
```diff
@@ -211,22 +214,45 @@ private void bindReaderToDeclaration( RuleBuildContext context, AccumulateDescr
...
         }
         return function;
     }
 
+    private String getFunctionName( RuleBuildContext context, AccumulateFunctionCallDescr fc, RuleConditionElement source, Map<String, Class<?>> declCls ) {
+        String functionName = fc.getFunction();
+        if (functionName.equals( "sum" )) {
+            Class<?> exprClass = MVELExprAnalyzer.getExpressionType( context, declCls, source, fc.getParams()[0] );
+            if (exprClass == int.class || exprClass == Integer.class) {
+                functionName = "sumI";
```
Comment (ge0ffrey) [inline drools-compiler/src/main/java/org/drools/compiler/rule/builder/dialect/java/JavaAccumulateBuilder.java:244]: For clarity, I believe they should have been named "sumInteger" etc, but I believe that ship has already sailed, didn't it?

File: drools-compiler/src/main/java/org/drools/compiler/rule/builder/dialect/java/JavaAccumulateBuilder.java [hunk 2]
```diff
@@ -211,22 +214,45 @@ private void bindReaderToDeclaration( RuleBuildContext context, AccumulateDescr
...
             context.addError( new DescrBuildError( accumDescr,
                                                    context.getRuleDescr(),
                                                    null,
-                                                   "Unknown accumulate function: '" + fc.getFunction() + "' on rule '" + context.getRuleDescr().getName() + "'. All accumulate functions must be registered before building a resource." ) );
+                                                   "Unknown accumulate function: '" + functionName + "' on rule '" + context.getRuleDescr().getName() + "'. All accumulate functions must be registered before building a resource." ) );
         }
         return function;
     }
 
+    private String getFunctionName( RuleBuildContext context, AccumulateFunctionCallDescr fc, RuleConditionElement source, Map<String, Class<?>> declCls ) {
```
Comment (ge0ffrey) [inline drools-compiler/src/main/java/org/drools/compiler/rule/builder/dialect/java/JavaAccumulateBuilder.java:239]: The getFunctionName doesn't deal with averageBD, maxDB, etc, so I suspect that the BigDecimalAverageAccumulateFunction. We or QA will need to build a unit test for very combination of functionName and type, just to be sure that they all work properly.

File: drools-core/src/main/java/org/drools/core/base/accumulators/BigDecimalSumAccumulateFunction.java [hunk 3]
```diff
@@ -16,96 +16,69 @@
...
         data.total = BigDecimal.ZERO;
     }
 
-    /* (non-Javadoc)
-     * @see org.kie.base.accumulators.AccumulateFunction#accumulate(java.lang.Object, java.lang.Object)
-     */
     public void accumulate(Serializable context,
                            Object value) {
         SumData data = (SumData) context;
         data.total = data.total.add( BigDecimal.valueOf( ( (Number) value ).doubleValue() ) );
```
Comment (ge0ffrey) [inline drools-core/src/main/java/org/drools/core/base/accumulators/BigDecimalSumAccumulateFunction.java]: total.add( BigDecimal.valueOf( ( (Number) value ).doubleValue() ) // Now my eyes bleed :)

Looks like we need to fix the existing implementation(s)

**Comment by ge0ffrey** [comment]
If you've seen any previous mails with stacktraces: those don't count - I had another local branch vs your branch name conflict issue locally. I deleted all those from the Github page, so only relevant comments are still on this github PR page.

There's still an issue: with a sum of integers, the call to IntegerSumAccumulateFunction.accumulate(Serializable context, Object value) actually delivers a Long value (which is nicely cast to an int, except in my drools fork (which fixes some bugs) where it triggers a ClassCastException).

It shouldn't be deliver an Long value. For example, in this rule:

```
rule "requiredCpuPowerTotal"
    when
        $computer : CloudComputer($cpuPower : cpuPower)
        $requiredCpuPowerTotal : Integer(this > $cpuPower) from accumulate(
            CloudProcess(
                computer == $computer,
                $requiredCpuPower : requiredCpuPower),
            sum($requiredCpuPower)
        )
    then
        scoreHolder.addHardConstraintMatch(kcontext, $cpuPower - $requiredCpuPowerTotal);
end
```

with this class

```
public class CloudProcess {
    private int requiredCpuPower;

    public int getRequiredCpuPower() {
        return requiredCpuPower;
    }
} 
```

That sum value should definitely be an `int` or `Integer`.

**Comment by ge0ffrey** [comment]
I've send in a pull request for your branch that fixes the BigDecimal rounding issue and exposes the Long casting issue: https://github.com/mariofusco/drools/pull/1

File: drools-core/src/main/java/org/drools/core/base/accumulators/BigDecimalSumAccumulateFunction.java (hunk 3 above)
Comment (ge0ffrey) [inline drools-core/src/main/java/org/drools/core/base/accumulators/BigDecimalSumAccumulateFunction.java]: I've send in a PR for that https://github.com/mariofusco/drools/pull/1

**Comment by mariofusco** [comment]
@ge0ffrey I merged your PR into mine but I still have a doubt: can we safely assume that an accumulator of Numbers of a given class will be exclusively fed by instances of that class? E.g. is it correct that an accumulator on BigDecimal eventually throws a CCE if for some reason is passed with a double? I must admit that at the moment I have no evidence of how this could happen, I just wanted to be sure that this is acceptable. Also what happens if it is passed with a null? I'm afraid this is far more probable and at the moment it will throw a NPE. What should we do in that case? Propagate the NPE? Considering it as a zero?

That said if I have understood correctly you're saying that this current implementation triggers an error in that requiredCpuPowerTotal rule? If so how could I reproduce this issue?

**Comment by ge0ffrey** [comment]
This PR fails, probably because of the ClassCastExceptions that are happening. Before we revert those casting and head back to doubleValue() (which is a rounding error when done some BigDecimals and even some longs), we might want to check if we can fix that and keep the casts.

About a double being passed to a BigDecimal accumulate: isn't that impossible because the compile time model guarantees that the parameter is of type BigDecimal? It's like saying that the method AtomicLong.addAndGet(long delta) should internally deal with delta not being a long but a double, no?
Yes, our unit tests now break with ClassCastException, but isn't that normal as they're trying to put ints in a Double accumulation etc? Maybe for backwards compatility we should allow the double accumulation to avoid the cast if it's not instanceof and fallback to doubleValue().

About nulls: if for backwards compatibility you want to keep null + null = null and null + 7 = 7 etc, I wouldn't mind, but otherwise I 'd just fail fast on that too.
Fail fast isn't bad: it's just saying we're not allowing you to do that now because we believe you'll only use it unintended in error - and once someone can deliver a use case that does really need it, we 'll put it in there in the next version. Josh Bloch's advice for any feature: **When in doubt, leave it out.** (especially edge cases are good fail fast candidates until proven useful).

Let's discuss Wednesday afternoon or Friday when I back in the home-office?

**Comment by mariofusco** [comment]
Ok, let's discuss this when you'll be back at home. For now I just want to give a look at that CCE you mentioned (and possibly fix it if I'll have time). How can I reproduce it?

**Comment by ge0ffrey** [comment]
To reproduce run AccumulateTest or any of the others that are failing on Jenkins.

**Comment by mariofusco** [comment]
Replaced by https://github.com/droolsjbpm/drools/pull/867

**Pull request closed without merging**